src/lib.rs

```diff
@@ -137,19 +151,389 @@ pub fn get_verifiable_credential_public_key(
+    params: AccountCredentialParameters,
+) -> Result<AccountCredentialResult, ConcordiumWalletCryptoError> {
+    serde_json::to_string(&params)
+        .context("cannot encode request object as JSON")
+        .and_then(|json| {
+            serde_json::from_str::<UnsignedCredentialInput>(&json)
+                .context("cannot decode request object into internal type")
+        })
+        .and_then(|input| {
+            create_unsigned_credential_v1_aux(input).context("cannot create identity")
+        })
+        .and_then(|res| {
+            serde_json::from_str::<AccountCredentialResult>(&res)
+                .context("cannot decode response object into result type")
+        })
+        .map_err(|e| ConcordiumWalletCryptoError::CallFailed {
+            call: "account_credential(...)".to_string(),
+            msg: format!("{:#}", e),
+        })
+}
+
+/// Duplicate of [`CredentialDeploymentDetails`] because that type only supports construction from JSON.
+/// It's only used internally and not exported.
+#[derive(Debug, Serialize)]
+pub struct CredentialDeploymentPayloadHashInput {
+    #[serde(rename = "expiry")]
+    pub expiry_unix: u64,
+    #[serde(rename = "unsignedCdi")]
+    pub credential: AccountCredential,
+}
+
+pub fn account_credential_deployment_hash_hex(
+    credential: AccountCredential,
+    expiry_unix: u64,
+) -> Result<String, ConcordiumWalletCryptoError> {
+    let input = CredentialDeploymentPayloadHashInput {
+        expiry_unix,
+        credential,
+    };
+    serde_json::to_string(&input)
+        .context("cannot encode request object as JSON")
+        .and_then(|json| {
+            serde_json::from_str::<CredentialDeploymentDetails>(&json)
+                .context("cannot decode request object into internal type")
+        })
+        .map(compute_credential_deployment_hash_to_sign)
+        .map_err(|e| ConcordiumWalletCryptoError::CallFailed {
+            call: "account_credential_deployment_hash_hex(...)".to_string(),
+            msg: format!("{:#}", e),
+        })
+}
+
+#[derive(Debug, Serialize)]
+pub struct SignedAccountCredential {
+    #[serde(rename = "unsignedCdi")]
+    pub credential: AccountCredential,
+    #[serde(rename = "signatures")]
+    pub signatures_hex: HashMap<KeyIndex, String>,
+}
+
+pub fn account_credential_deployment_signed_payload_hex(
+    credential: SignedAccountCredential,
+) -> Result<String, ConcordiumWalletCryptoError> {
+    serde_json::to_string(&credential)
+        .context("cannot encode request object as JSON")
+        .and_then(|json| {
+            serde_json::from_str::<CredentialDeploymentPayload>(&json)
+                .context("cannot decode request object into internal type")
+        })
+        .map(serialize_credential_deployment_payload)
+        .map_err(|e| ConcordiumWalletCryptoError::CallFailed {
+            call: "account_credential_deployment_signed_payload_hex(...)".to_string(),
+            msg: format!("{:#}", e),
+        })
+}
```

Comment:
As already answered in other comments, the reason is UniFFI. And the reason why I want to use that is that it allows us to generate *a lot* of boilerplate on the Swift side. Not just types but also very error prone scaffolding around the FFI calls. The functions are exposed simply as regular Swift functions, including proper exceptions.

And because UniFFI supports several other languages, this is all reusable.

I'm open to doing things in a simpler way if that is possible. Duplication is definitely not ideal, but as I see it, the alternative is implementing `UniffiCustomTypeConverter` for the existing types which I think would be much more complex. I do not want to do all the things that UniFFI does for me by hand just to avoid duplicating some structs.

Reply:
Ok.

I am fine with using UniFFI since you spent all this time on it now. I don't think the rationale about "other languages" is relevant here though since the supported ones we already actually have and we won't be rewriting things.

But I still don't understand why you are not using UniFFI code generation as I read it should be used. The Rust types you are defining manually.

Reply:
Discussed offline.